task-context.md

```diff
@@ -0,0 +1,193 @@
+- Feature Name: task-context
+- Start Date: 2018-01-22
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+Learnability, reasoning, debuggability and `no_std`-compatibility improvements achieved by
+making task wakeup handles into explicit arguments.
+
+# Motivation
+[motivation]: #motivation
+
+## Learnability
+
+The `futures` task wakeup mechanism is [a regular source of confusion for new
+users][explicit task issue]. Looking at the signature of `Future::poll`, it's
+hard to tell how `Future`s work:
+
+```rust
+fn poll(&mut self) -> Poll<Self::Item, Self::Error>;
+```
+
+`poll` takes no arguments, and returns either `Ready`, `NotReady`, or an error.
+From the signature, it looks as though the only way to complete a `Future` is
+to continually call `poll` in a loop until it returns `Ready`. This is
+incorrect.
+
+Furthermore, when implementing `Future`, it's necessary to schedule a wakeup
+when you return `NotReady`, or else you might never be `poll`ed again. The
+trick to the API is that there is an implicit `Task` argument to `Future::poll`
+which is passed using thread-local storage. This `Task` argument must be
+`notify`d by the `Future` in order for the task to awaken and `poll` again.
+It's essential to remember to do this, and to ensure that the right schedulings
+occur on every code path.
+
+[explicit task issue]: https://github.com/alexcrichton/futures-rs/issues/129
+
+## Reasoning about code
+
+The implicit `Task` argument makes it difficult to tell which functions
+aside from `poll` will use the `Task` handle to schedule a wakeup. That has a
+few implications.
```

Comment:
You can tell which functions will use the `Task` handle because they return `Async`.

Today, you can look at any function in the `h2` library and know if it requires a task context or not.

Reply:
You can't tell that easily from the *call site*, which is the key difference in code reasoning.

Reply:
There are also functions like `need_read` which don't return `Async` but do rely on an implicit `Task`.

Reply:
> You can't tell that easily from the call site, which is the key difference in code reasoning.

But you can, it's determined by the return value. Unless you mean when people rely on type inference for the return value and don't actually check that it's needed?

Reply:
I believe I've seen code use `Async` as a "general purpose" return value to indicate readiness, without necessarily touching a `Task` internally.  Perhaps this practice should be discouraged if the "Async as a return value means we touched a Task" convention is to take hold.

Reply:
But you can't *easily*.